gomock/internal/validate/validate.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validate
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+)
+
+// InputAndOutputSig compares the argument and return signatures of actualFunc
+// against expectedFunc. It returns an error unless everything matches.
+func InputAndOutputSig(actualFunc, expectedFunc reflect.Type) error {
+	if err := InputSig(actualFunc, expectedFunc); err != nil {
+		return err
+	}
+
+	if err := outputSig(actualFunc, expectedFunc); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// InputSig compares the argument signatures of actualFunc
+// against expectedFunc. It returns an error unless everything matches.
+func InputSig(actualFunc, expectedFunc reflect.Type) error {
+	// check number of arguments and type of each argument
+	if actualFunc.NumIn() != expectedFunc.NumIn() {
+		return fmt.Errorf(
+			"expected function to have %d arguments not %d",
+			expectedFunc.NumIn(), actualFunc.NumIn())
+	}
+
+	lastIdx := expectedFunc.NumIn()
+
+	// If the function has a variadic argument validate that one first so that
+	// we aren't checking for it while we iterate over the other args
+	if expectedFunc.IsVariadic() {
+		if ok := variadicArg(lastIdx, actualFunc, expectedFunc); !ok {
+			i := lastIdx - 1
+			return fmt.Errorf(
+				"expected function to have"+
+					" arg of type %v at position %d"+
+					" not type %v",
+				expectedFunc.In(i), i, actualFunc.In(i),
+			)
+		}
+
+		lastIdx--
+	}
+
+	for i := 0; i < lastIdx; i++ {
+		expectedArg := expectedFunc.In(i)
+		actualArg := actualFunc.In(i)
+
+		if err := arg(actualArg, expectedArg); err != nil {
+			return fmt.Errorf("input argument at %d: %s", i, err)
+		}
+	}
+
+	return nil
+}
+
+func outputSig(actualFunc, expectedFunc reflect.Type) error {
+	// check number of return vals and type of each val
+	if actualFunc.NumOut() != expectedFunc.NumOut() {
+		return fmt.Errorf(
+			"expected function to have %d return vals not %d",
+			expectedFunc.NumOut(), actualFunc.NumOut())
+	}
+
+	for i := 0; i < expectedFunc.NumOut(); i++ {
+		expectedArg := expectedFunc.Out(i)
+		actualArg := actualFunc.Out(i)
+
+		if err := arg(actualArg, expectedArg); err != nil {
+			return errors.Wrapf(err, "return argument at %d", i)
```

Comment:
Do we need to wrap the errors? Could we append the old error as details? Or if we do want to wrap maybe we should consider `https://github.com/golang/xerrors/` instead as that is the api in go1.13. Thoughts?

Reply:
I don't think it's necessary to wrap them; I actually did it out of habit and raised the same question in chat. I recently read a great posts about how the errors you return are part of your public API, wrapping errors expands the surface of that API.

I'm happy to go with whatever you prefer.

Reply:
I would say for now let just not wrap since there is not a strong case for it. This has the benefit of pushing one less dependency on our users too.

Reply:
Done